distribution/lib/Standard/Base/0.0.0-dev/src/Data/List.enso

```diff
@@ -162,26 +166,31 @@ type List
     not_empty : Boolean
     not_empty self = self.is_empty.not
 
-    ## Selects all elements of this list which satisfy a predicate.
+    ## Selects all elements of this list which satisfy a condition.
 
        Arguments:
-       - predicate: A function that takes a list element and returns a boolean
-         value that says whether that value satisfies the conditions of the
-         function.
+       - filter: The filter to apply to the list. It can either be an instance
+         of `Filter_Condition` or a predicate taking a value and returning a
+         boolean value indicating whether the corresponding element should be
+         kept or not.
 
        > Example
          Selecting all elements that are less than 3.
 
              import Standard.Examples
 
              example_filter = Examples.list.filter (< 3)
-    filter : (Any -> Boolean) -> List
-    filter self predicate =
-        case self of
-            Cons a b ->
-                rest = b.filter predicate
-                if predicate a then Cons a rest else rest
-            Nil -> Nil
+             example_filter = Examples.list.filter (Filter_Condition.Less than=3)
+    filter : (Filter_Condition | (Any -> Boolean)) -> Vector Any
+    filter self filter = case filter of
+        _ : Filter_Condition -> self.filter (Filter_Condition_Module.to_predicate filter)
+        predicate : Function ->
+            go_filter list = case list of
+                Cons h t ->
+                    rest = go_filter t
+                    if predicate h then Cons h rest else rest
+                Nil -> Nil
+            go_filter self
```

Comment:
I left the logic for the filter operation as-is. But it is slightly concerning to me that it is _not_ tail-recursive. It may overflow for very big lists.

@jdunkerley do you think we should add a test with a very big list to check for the overflow? Seems like this should work regardless of size. We can try to rewrite it using for example fold+reverse to regain tail recursion.

Reply:
we need to redo `List` as a built in - we should cover it there.